include/seastar/coroutine/exception.hh

```diff
@@ -23,18 +23,47 @@
 
 #include <seastar/core/coroutine.hh>
 
-namespace seastar::coroutine {
+namespace seastar {
+
+namespace internal {
+
+struct exception_awaiter {
+    std::exception_ptr eptr;
+
+    explicit exception_awaiter(std::exception_ptr&& eptr) : eptr(std::move(eptr)) {}
+
+    exception_awaiter(const exception_awaiter&) = delete;
+    exception_awaiter(exception_awaiter&&) = delete;
+
+    bool await_ready() const noexcept {
+        return false;
+    }
+
+    template<typename U>
+    void await_suspend(SEASTAR_INTERNAL_COROUTINE_NAMESPACE::coroutine_handle<U> hndl) noexcept {
+        hndl.promise().set_exception(std::move(eptr));
+        hndl.destroy();
```

Comment:
Nice hack. Please adjust the test to confirm that destructors in the coroutine are executed. They absolutely should, but I saw no hint of that in the documentation.

Reply:
I modified the tests so that the coroutines allocate an object on the stack which increments a counter in constructor and decrements it on destruction. When I removed `hndl.destroy()`, the assertion `count == 0` failed and ASan reported a memory leak, so it suggests that the destructors are run by `coroutine_handle::destroy`.

I also found _some_ version of the C++20 draft and it mentions that `coroutine_handle::destroy` runs destructors of automatic variables in the coroutine state: [dcl.fct.def.coroutine#11](https://timsong-cpp.github.io/cppwp/n4861/dcl.fct.def.coroutine#11). After `await_ready()` returns false the coroutine is considered suspended, so calling `destroy` here shouldn't be considered as undefined behavior ([expr.await#5.1](https://timsong-cpp.github.io/cppwp/n4861/expr.await#5.1)).

Reply:
The idea for the "hack" was heavily inspired by this article: https://ladnir.github.io/blog/2020/03/31/result.html

I'm wondering if it would be possible to extend the idea and devise a wrapper over Seastar futures which allows co_awaiting on futures and propagates exceptions without throwing. Its main drawback would be that it wouldn't honor try..catch blocks, just like coroutine::exception, but it could be useful most of the time. Would something like this accepted?

Reply:
Sounds good to me, as a heavy user of `return make_exception_future<T>()`